src/werkzeug/formparser.py

```diff
@@ -423,156 +411,322 @@ def validate_boundary(self, boundary):
-                # this is usually a newline delimiter.
-                if buf:
-                    yield _cont, buf
-                    buf = b""
-
-                # If the line ends with windows CRLF we write everything except
-                # the last two bytes.  In all other cases however we write
-                # everything except the last byte.  If it was a newline, that's
-                # fine, otherwise it does not matter because we will write it
-                # the next iteration.  this ensures we do not write the
-                # final newline into the stream.  That way we do not have to
-                # truncate the stream.  However we do have to make sure that
-                # if something else than a newline is in there we write it
-                # out.
-                if line[-2:] == b"\r\n":
-                    buf = b"\r\n"
-                    cutoff = -2
+        def _state_pre_term(self, line):
+            """State for the very beginning of the stream, before any content.
+            Eats empty lines until it finds a boundary line."""
+            if not line:
+                raise ValueError("Unexpected end of file")
+                return self._state_pre_term
+            line = line.rstrip(b"\r\n")
+            if not line:
+                return self._state_pre_term
+            if line == self._last_part:
+                return self._state_done
+            elif line == self._next_part:
+                self._headers = []
+                return self._state_headers
+            raise ValueError("Expected boundary at start of multipart data")
+
+        def _state_headers(self, line):
+            """State for field headers. They are parsed and left in
+            ``self._headers``."""
+            if line is None:
+                raise ValueError("Unexpected end of file during headers")
+            line = to_native(line)
+            line, line_terminated = _line_parse(line)
+            if not line_terminated:
+                raise ValueError("Unexpected end of line in multipart header")
+            if not line:
+                self._headers = Headers(self._headers)
+                return self._start_content()
+            if line[0] in " \t" and self._headers:
+                key, value = self._headers[-1]
+                self._headers[-1] = (key, value + "\n " + line[1:])
+            else:
+                parts = line.split(":", 1)
+                if len(parts) == 2:
+                    self._headers.append((parts[0].strip(), parts[1].strip()))
                 else:
-                    buf = line[-1:]
-                    cutoff = -1
-                yield _cont, line[:cutoff]
+                    raise ValueError("Malformed header")
+            return self._state_headers
+
+        def feed(self, lines):
+            self._output = []
+            s = self._state
+            for line in lines:
+                s = s(line)
```

Comment:
Re-assigning to this function works but is a little confusing. Consider having an enum like

```
class ParseState(enum.Enum):
  STATE_DONE, STATE_OUTPUT, ... = range(2)  # or however many states
```

Then, here, there can just be a mapping like

```
transition_functions = {
  ParseState.STATE_DONE: self._state_done,
  ...
}
```

Each transition function would ideally be a generator, but that might be tough since the transitions also need to indicate what the next state should be. Conceivably, this could look like generators that yield tuples like `(some_output, new_state)` or `(some_output, None)`.